service/src/main/java/com/epam/esm/service/service/certificate/CertificateServiceImpl.java

```diff
@@ -0,0 +1,286 @@
+                }
+                break;
+        }
+    }
+
+    private void addToFinder(Consumer<String> consumer, String value) {
+        if (StringUtils.isNotEmpty(value)) {
+            consumer.accept(decodeParam(value));
+        }
+    }
+
+    private void addToFinder(Consumer<String> consumer, List<String> list) {
+        if (CollectionUtils.isNotEmpty(list)) {
+            addToFinder(consumer, list.get(0));
+        }
+    }
+
+    @Override
+    public List<Certificate> read(MultiValueMap<String, String> params) throws BadRequestException, NotFoundException {
+        CertificateFinder finder = new CertificateFinder();
+        for (String key : params.keySet()) {
+            try {
+                if (CollectionUtils.isEmpty(params.get(key))) {
+                    throw new BadRequestException("Empty parameter!", ErrorCodes.CERTIFICATE_BAD_REQUEST);
+                }
+                if (key.contains("sort")) {
+                    finder.sortBy(CertificateSortingParameters.getEntryByParameter(key).getValue(),
+                            SortDirection.parseAscDesc(params.get(key).get(0)));
+                } else if (PaginationParameters.contains(key)) {
+                    parsePaginationParameter(finder, key, params.get(key));
+                } else {
+                    parseFindParameter(finder, key, params.get(key));
+                }
+            } catch (IllegalArgumentException e) {
+                throw new BadRequestException(e, ErrorCodes.CERTIFICATE_BAD_REQUEST);
+            }
+        }
+        return readBy(finder);
+    }
+
+    @Transactional
+    @Override
+    public Certificate patch(Certificate certificate) throws ValidationException, BadRequestException {
+        if (certificate.getId() <= 0) {
+            throw new BadRequestException("Wrong certificate id!", ErrorCodes.CERTIFICATE_BAD_REQUEST);
+        }
+
+        Certificate oldCertificate = dao.read(certificate.getId());
+
+        if (oldCertificate == null) {
+            throw new BadRequestException("Wrong certificate id!", ErrorCodes.CERTIFICATE_BAD_REQUEST);
+        }
+        if (certificate.getPrice() != null
+                && certificate.getPrice().compareTo(BigDecimal.ZERO) > 0 ) {
+            oldCertificate.setPrice(certificate.getPrice());
+        }
+        if (certificate.getDescription() != null) {
+            oldCertificate.setDescription(certificate.getDescription());
+        }
+        if (certificate.getDuration() != 0) {
+            oldCertificate.setDuration(certificate.getDuration());
+        }
+        if (certificate.getName() != null) {
+            oldCertificate.setName(certificate.getName());
+        }
+        return update(oldCertificate);
+    }
+
+    @Transactional
+    @Override
+    public void addCertificateTag(Certificate certificate, int tagId) throws ServiceException,
+            BadRequestException, ValidationException {
+        Optional<Certificate> certificateOptional;
```

Comment:
the same as before, there is no need to separate variable declaration and initialization

Reply:
Done